File: CS/GrowthWare.Framework/Model/Profiles/Base/MBaseSecurity.cs
using GrowthWare.Framework.Model.Enumerations;
using GrowthWare.Framework.Model.Profiles.Interfaces;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GrowthWare.Framework.Model.Profiles.Base
{
    /// <summary>
    /// The MBaseSecurity is a abstract class that when inherited will add 4 types of roles
    /// to your class/object.  After you have inherited the class pass a data row to the SecurityInit sub 
    /// to populate the roles.
    /// </summary>
    /// <remarks>
    /// Currently there are 4 permission roles and they are Add, Edit, View, Delete.  
    /// If you would like to extend this class do so by inheriting this class and adding your
    /// own types of roles say Moderate if your writing some sort of formum type of class.
    /// Any of the objects you create should now inherit your class and will now have all of the 
    /// roles from this class as well as the ones for yours.
    ///</remarks>
    [Serializable(), CLSCompliant(true)]
    public abstract class MBaseSecurity : MProfile, IMSecurityInfo
    {
        private string[] m_AddRoles;
        private string[] m_DeleteRoles;
        private string[] m_EditRoles;
        private string[] m_ViewRoles;
        private string m_PermissionColumn = "PERMISSIONS_SEQ_ID";

        private string m_RoleColumn = "ROLE";
        /// <summary>
        /// Init orverloads and calles mybase.init to will populate the Add, Delete, Edit, and View role properties.
        /// </summary>
        /// <param name="dataRowSecurity">A data row that must contain two columns ("PERMISSIONS_SEQ_ID","ROLE")</param>
        /// <param name="dataRowMain">A data row that contains base information</param>
        /// <remarks></remarks>
        protected void Init(DataRow dataRowMain, DataRow[] dataRowSecurity)
        {
            base.Initialize(dataRowMain);
            m_AddRoles = SplitRoles(dataRowSecurity, RoleType.AddRole);
            m_DeleteRoles = SplitRoles(dataRowSecurity, RoleType.DeleteRole);
            m_EditRoles = SplitRoles(dataRowSecurity, RoleType.EditRole);
            m_ViewRoles = SplitRoles(dataRowSecurity, RoleType.ViewRole);
        }

        public string[] AddRoles
        {
            get { return m_AddRoles; }
        }

        public string[] DeleteRoles
        {
            get { return m_DeleteRoles; }
        }

        public string[] EditRoles
        {
            get { return m_EditRoles; }
        }

        public string[] ViewRoles
        {
            get { return m_ViewRoles; }
        }

        public string PermissionColumn
        {
            get { return m_PermissionColumn; }
            set { m_PermissionColumn = value.Trim(); }
        }

        public string RoleColumn
        {
            get { return m_RoleColumn; }
            set { m_RoleColumn = value.Trim(); }
        }

        protected string[] SplitRoles(DataRow[] allRoles, RoleType moduleRoleType)
        {
            ArrayList colRoles = new ArrayList();
            DataRow row = null;
            foreach (DataRow row_loopVariable in allRoles)
            {
                row = row_loopVariable;
                if (!Convert.IsDBNull(row[m_PermissionColumn]))
                {
                    if ((RoleType)row[m_PermissionColumn] == moduleRoleType)
                    {
                        colRoles.Add(row[m_RoleColumn]);
                    }
                }
            }
            return (string[])colRoles.ToArray(typeof(string));
        }
    }
}

File: VB/GrowthWare.Framework/Model/Profiles/Base/MBaseSecurity.vb
Imports GrowthWare.Framework.Model.Enumerations
Imports GrowthWare.Framework.Model.Profiles.Interfaces

Namespace Model.Profiles.Base
    Public MustInherit Class MBaseSecurity
        Inherits MProfile
        Implements IMSecurityInfo

        Private m_AddRoles() As String
        Private m_DeleteRoles() As String
        Private m_EditRoles() As String
        Private m_ViewRoles() As String
        Private m_PermissionColumn As String = "PERMISSIONS_SEQ_ID"
        Private m_RoleColumn As String = "ROLE"

        ''' <summary>
        ''' Init orverloads and calles mybase.init to will populate the Add, Delete, Edit, and View role properties.
        ''' </summary>
        ''' <param name="drowSecurity">A data row that must contain two columns ("PERMISSIONS_SEQ_ID","ROLE")</param>
        ''' <remarks></remarks>
        Protected Overloads Sub Initialize(ByVal dataRow As DataRow, ByRef drowSecurity() As DataRow)
            MyBase.Initialize(dataRow)
            m_AddRoles = SplitRoles(drowSecurity, RoleType.AddRole)
            m_DeleteRoles = SplitRoles(drowSecurity, RoleType.DeleteRole)
            m_EditRoles = SplitRoles(drowSecurity, RoleType.EditRole)
            m_ViewRoles = SplitRoles(drowSecurity, RoleType.ViewRole)
        End Sub

        Public ReadOnly Property AddRoles() As String() Implements IMSecurityInfo.AddRoles
            Get
                Return m_AddRoles
            End Get
        End Property

        Public ReadOnly Property DeleteRoles() As String() Implements IMSecurityInfo.DeleteRoles
            Get
                Return m_DeleteRoles
            End Get
        End Property

        Public ReadOnly Property EditRoles() As String() Implements IMSecurityInfo.EditRoles
            Get
                Return m_EditRoles
            End Get
        End Property

        Public ReadOnly Property ViewRoles() As String() Implements IMSecurityInfo.ViewRoles
            Get
                Return m_ViewRoles
            End Get
        End Property

        Public Property PermissionColumn() As String
            Get
                Return m_PermissionColumn
            End Get
            Set(ByVal value As String)
                m_PermissionColumn = value.Trim
            End Set
        End Property

        Public Property RoleColumn() As String
            Get
                Return m_RoleColumn
            End Get
            Set(ByVal value As String)
                m_RoleColumn = value.Trim
            End Set
        End Property

        Protected Function SplitRoles(ByVal allRoles() As DataRow, ByVal moduleRoleType As RoleType) As String()
            Dim colRoles As New ArrayList
            Dim row As DataRow
            For Each row In allRoles
                If Not IsDBNull(row(m_PermissionColumn)) Then
                    If CType(row(m_PermissionColumn), RoleType) = moduleRoleType Then
                        colRoles.Add(row(m_RoleColumn))
                    End If
                End If
            Next row
            Return CType(colRoles.ToArray(GetType(String)), String())
        End Function
    End Class
End Namespace
